flowmachine/flowmachine/core/subscriber_subset.py

```diff
@@ -0,0 +1,186 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from abc import abstractmethod
+from sqlalchemy.sql import ClauseElement, select, text, column
+from .query import Query
+
+
+class SubscriberSubsetBase(Query):
+    """
+    Base class for the different types of subscriber subsets.
+    """
+
+    @property
+    @abstractmethod
+    def is_proper_subset(self):
+        raise NotImplementedError(
+            f"Class {self.__class__.__name__} does not implement 'is_proper_subset'"
+        )
+
+    @abstractmethod
+    def apply_subset(self, sql, *, subscriber_identifier):
+        raise NotImplementedError(
+            f"Class {self.__class__.__name__} does not implement 'apply_subset'"
+        )
+
+    def _get_query_attrs_for_dependency_graph(self, analyse=False):
+        # This is a stub implementation of this internal method.
+        # It is needed because SubscriberSubsetBase currently
+        # inherits from flowmachine.Query, so we implement just
+        # enough to ensure Query.dependency_graph() doesn't break.
+        attrs = {}
+        attrs["name"] = self.__class__.__name__
+        attrs["stored"] = "N/A"
+        attrs["cost"] = "N/A"
+        attrs["runtime"] = "N/A"
+        return attrs
+
+
+class AllSubscribers(SubscriberSubsetBase):
```

Comment:
I think the existence of this class slightly confusing, I think we should not be calling down below
   
```
select_stmt = self.subscriber_subset.apply_subset(
            select_stmt, subscriber_identifier=self.subscriber_identifier
        ) 
```

when we are not extracting a subset. I would rather go for an if-statement like `if subset: ...`

Reply:
I can see where you're coming from, but the existence of this class is actually crucial to reduce / eliminate coupling with other parts of the code. Let me try to explain why.

One of the reasons why the current code base is so entangled is because certain classes apply logic or behaviour _on behalf of_ other classes when they shouldn't need to know anything about how things are implemented. A lot of the `if`/`elif`/`else` statements in the code currently have this purpose (especially where they have the form`if isinstance(...)`).

If we omit the `AllSubscribers` case here then we suddenly have a special case that needs to be treated _elsewhere_ depending on whether we do or do not have a proper subset. This check might only happen once (for example in `EventTableSubset`), but once other classes are also interested in subsetting, all of a sudden this if/else statement is duplicated in multiple different places. If you later change anything about the implementation of the subset classes which affects how this check works you end up having to do "shotgun surgery" and change it in all those different places.

By contrast, if you _do_ have an `AllSubscribers` class then there are no ripple effects because all types of subsets are treated exactly the same. So the only thing that any external code needs to know is how to apply a subset (by calling the `.apply_subset()` method of the subset class, no matter which of the multiple different cases it is).

I agree that it feels strange when you first see it, but it's actually a powerful technique to eliminate dependencies. It's also in the spirit of true object-oriented programming (rather than the procedural-style programming inside classes that we have in most places) where instead of _asking_ for information from an object (here: a subset) and then applying different actions depending on the answer you simply _tell_ it what to do (here: apply a subset, whatever it is) and trust that it knows how to do it.

I'm happy to dig up more resources and examples of why this is useful if you're interested, but all I can say at the moment is that using this technique in more places will allow us to massively reduce coupling and dependencies in the code.

Reply:
Thinking about it, maybe we can reduce the confusion by renaming the method `.apply_subset()` to `.apply_subset_if_needed()`? This would make it clearer that the "no subsetting needed" case is intentionally included.

Reply:
Got it. Very clear explanation. I am indifferent about renaming, but feel more strongly about `apply_subset` just because it is shorter. At the end of the day, selecting all is a special subset case.